### PR TITLE
Updated bulk importer to improve performance

### PR DESCRIPTION
**Pull request by allouis**
no-issue

The main changes here are:
- Refactor to remove global state, hopefully easier to maintain
- Do everything we can do in parallel
- Use a Leaky Bucket to make requests to Stripe API as fast as possible
- Do not add members that we know we will delete (e.g. ones with stripe
  customers when stripe is not enabled)
- Remove as many loops as possible, e.g. combine filter and map into a
  single reduce
- Handle live & test stripe rate limits
- Remove support for Complimentary plans (it has to be reworked anyway)

### CODE REVIEW

**Comment by allouis** [comment]
@naz I've used a PR for this & requested your review on this because it's quite a big changeset, and I didn't wanna push to master and screw you over.

Basic idea here is:
1. Do as many things in parallel as we can
2. Use the Leaky Bucket to request as fast as possible from Stripe API

File: core/server/api/canary/members.js [hunk 1]
```diff
@@ -604,7 +604,10 @@ module.exports = {
                     invalid,
                     createdBy
                 });
-            }).then(() => {
+            }).then((result) => {
+                invalid.errors = invalid.errors.concat(result.invalid.errors);
```
Comment (naz) [inline core/server/api/canary/members.js:608]: Yesss! +100 points for getting rid of those 2 extra parameters!

File: core/server/services/members/importer/index.js [hunk 2]
```diff
@@ -80,150 +179,92 @@ const doImport = async ({members, allLabelModels, importSetLabels, imported, inv
...
-                .where('id', id)
-                .del();
-
-            if (deletedMembersCount) {
-                imported.count -= deletedMembersCount;
-                invalid.count += deletedMembersCount;
-            }
-        };
+async function fetchStripeCustomers(stripeCustomersToInsert) {
+    const LeakyBucket = require('leaky-bucket');
```
Comment (naz) [inline core/server/services/members/importer/index.js:225]: Awesome one. Definitely a thing to rely upon stead of testing at which concurrency count things break :+1:

File: core/server/services/members/importer/index.js [hunk 3]
```diff
@@ -4,50 +4,153 @@ const ObjectId = require('bson-objectid');
...
+    });
+
+    return result;
 };
 
-function serializeMemberLabels(labels) {
-    if (_.isString(labels)) {
-        if (labels === '') {
-            return [];
+const CHUNK_SIZE = 5000;
```
Comment (naz) [inline core/server/services/members/importer/index.js:76]: Would leave a note here about the CHUNK_SIZE value - we still need to test what's the optimal value for it is. Based on [these RoTs](http://mysql.rjweb.org/doc.php/ricksrots) the number should be between 100 and 1000.

File: core/server/services/members/importer/index.js [hunk 4]
```diff
@@ -4,50 +4,153 @@ const ObjectId = require('bson-objectid');
...
+    const stripeCustomersToInsert = [];
+    const labelAssociationsToInsert = [];
+
+    members.forEach(function (member) {
+        if (member.stripe_customer_id && !stripeIsConnected) {
+            invalidMembers.push(member);
+            return;
+        }
+
+        // @TODO This is expensive, maybe we can just error if we get shoddy data?
```
Comment (naz) [inline core/server/services/members/importer/index.js:148]: I'd even be up for dropping this whole block because it was added to handle a bug that happened in members export [pre version 3.6](https://github.com/TryGhost/Ghost/blob/78911257f407da311cf6a62bc148a51e5de4a4fb/core/server/api/canary/members.js#L16-L19) - limited time window. This is just my opinion, would need verification!

**Comment by naz** [comment]
@allouis I do like the direction and think we should merge these once the import works for for all needed scenarios. The missing bits I have in mind for these to be mergeable: members_stripe_customers_subscriptions records insertion and the `comped` flag handling. This is an awesome improvement!

File: core/server/api/canary/members.js (hunk 1 above)
Comment (allouis) [reply core/server/api/canary/members.js:608]: :hugs:

File: core/server/services/members/importer/index.js (hunk 2 above)
Comment (allouis) [reply core/server/services/members/importer/index.js:225]: Yeah! The implementation is pretty big though, has a lot of unnecessary deps :confused: The newest version looks cool, but needs to be run with an experimental flag :weary:  Might be worth writing our own, or finding a better library in the future.

File: core/server/services/members/importer/index.js (hunk 4 above)
Comment (allouis) [reply core/server/services/members/importer/index.js:148]: Okay, maybe one for @matthanley to take a look at?

TL;DR pre 3.6 we would export the string 'undefined' for undefined values, and we now have to loop through every key in every object to handle that - which is expensive, especially for large imports.

Would you consider dropping the support for that?  We could provide a command/tool to clean the data up locally if necessary OR we could offload that work to Ghost-Admin, that way the operation would not affect the request cycle

**Comment by naz** [comment]
@allouis should we merge this into master as is? Asking because that would help fixing the bug with rate-limit error in "validate import" :wink:
